cts.json

```diff
@@ -3819,6 +3819,58 @@
       "selector": "$[-1:-10:-231584178474632390847141970017375815706539969331281128078915168015826259279872]",
       "invalid_selector": true
     },
+    {
+      "name": "functions, arguments, parenthesized expression",
+      "selector": "$[?match((@.a), ('a.*'))]",
+      "document": [
+        {
+          "a": "ab"
+        },
+        {
+          "a": "ba"
+        }
+      ],
+      "result": [
+        {
+          "a": "ab"
+        }
+      ]
+    },
+    {
+      "name": "functions, arguments, function expression",
+      "selector": "$.values[?match(@.a, value($..['regex'])]",
+      "document": {
+        "regex": "a.*",
+        "values": [
+          {
+            "a": "ab"
+          },
+          {
+            "a": "ba"
+          }
+        ]
+      },
+      "result": [
+        {
+          "a": "ab"
+        }
+      ]
+    },
+    {
+      "name": "functions, arguments, comparison expression",
+      "selector": "$[?length(@.a==@.b)>2]",
```

Comment:
I'm not sure what this one is trying to do.  `@.a==@.b` yields a `LogicalType`, but `length()` takes a `ValueType`.  This path is semantically invalid (syntactically valid, though).  It shoudl error.

Reply:
The intention was to test that a comparison expression is a valid function argument, within the confines of the standard function extensions. 

I see that if `LogicalTrue` and `LogicalFalse` are not equal to JSON `true` and `false`, then the result of `@.a==@.b` would not be a `ValueType`, making the query invalid.

If `LogicalType` can be used in place of a JSON Boolean (implicitly converting it to a `ValueType`?), I would expect `length()` to return `Nothing`, and the comparison between `Nothing` and `2` to result in `LogicalFalse`.

I'm not sure, and I can always define some mock functions that do accepts a `LogicalType` parameter, outside of the CTS.

Reply:
We opted to separate the result of a boolean expression from the JSON literals `true` and `false` and recognize that the JSON literals are just tokens that don't carry any meaning.  There's good reason to do this, and I'll let you peruse the issues for more, but the main was that if we didn't separate these concepts, then test expressions would be ambiguous.

For example, consider the complete expression `@.foo`.  Is this testing whether `foo` exists, or whether `foo` contains a JSON `true`?  Unless we are explicit that the JSON `true` is separate from logical true, this can't be determined.

I expect that this test case should generate an error.  @glyn can you confirm, please?  What's the rule for syntactically correct but semantically incorrect paths?

Reply:
Thank you, @gregsdennis , that makes sense. I've provisionally marked that test case as invalid and given it a more relevant name.

[Section 2.4.9](https://datatracker.ietf.org/doc/html/draft-ietf-jsonpath-base-20#section-2.4.9) has some examples that cover of this kind of thing. I had missed that before.